### PR TITLE
Add support for TRI7 geometric elements

### PR DESCRIPTION
**Pull request by nmnobre**
## Reason
The objective is to implement TET14s next and, ultimately, to add support for the Raviart-Thomas family on more than one geometric element. As is, we would only be able to support HEX27s, which is fine... but precludes comparisons between TETs and HEXs which might be important for performance or numerical comparison purposes.

## Design
Initially, I thought I could implement TET14s and TRI7s together in one PR, but then I realised that there are enough points for discussion on the TRI7s alone, so I decided to split them.

1) Throughout moose, it's common to see "second order" being referred to as "high order". This raises the question if there's an interest in going beyond second order in the first place, specially since the "third order" enabled here is only P2+ really, i.e. these are no TRI10s... I've done some renaming and named things "third_order" or "cubic" but I'd like your opinion on this. :)

2) I haven't added support for TRI7s in the spiral annular mesh toolset, mostly due to its usage of `MeshBase::all_second_order()` for TRI6s. This would require the creation of a `MeshBase::all_third_order()` but, again, this wouldn't really be true to its name, so I thought we better discuss it before I did anything more profound.

3) You'll notice that in the `auxkernels/projection_aux.project/2d_cubic_tri` test, I'm using `ICs/active='quadratic_elem quadratic_nodal'` which will probably sound weird. But these functions are already cubic polynomials, is this on purpose?

https://github.com/idaholab/moose/blob/625f4ea51c26d127096438b7c0755e4ec1912fef/test/tests/auxkernels/projection_aux/2d.i#L35-L44

4) In `AdvancedExtruderGenerator::generate()` I'm building PRISM21s, so we might need to add formal support for these?

## Impact
By itself, this PR should enable P2 with bubble (aka P2+) simulations on 2d triangular meshes.
This will also enable the implementation of TET14s and Raviart-Thomas (very) shortly after its merged.

##

Cheers :),
-Nuno (cc @karthichockalingam)

### CODE REVIEW

**Comment by roystgnr** [comment]
Anybody know what's going on with the failed CI builds?  I'm just getting an interminable "Waiting for results" from the Conda build (Linux) and a surely-unrelated SUPERLU_DIST configure failure from the Clang build.

**Comment by lindsayad** [comment]
I believe it's https://moosedevelopers.slack.com/archives/C011ZQT8ABD/p1696872532952309?thread_ts=1696863034.872579&cid=C011ZQT8ABD

**Comment by GiudGiud** [comment]
CI is ded rn because github.inl.gov was down

File: test/tests/misc/max_var_n_dofs_per_elem/tests
Comment (GiudGiud) [inline test/tests/misc/max_var_n_dofs_per_elem/tests:50]: are we testing the postprocessor here?
If it should be a csvdiff not an exodiff 

this is probably because it s a really old test

File: framework/src/meshgenerators/AdvancedExtruderGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/AdvancedExtruderGenerator.C:17]: this reminds me that most of the changes you made for the other generators for TRI7 could also be done for PRISM21.

no need to if you dont need to use those

**Review by GiudGiud — COMMENTED** [review]
looks good to me but @roystgnr is more knowledgeable

File: test/tests/misc/max_var_n_dofs_per_elem/tests
Comment (nmnobre) [reply test/tests/misc/max_var_n_dofs_per_elem/tests:50]: Done for the existing tests in #25727.
I'll rebase and apply the same changes here once that's merged. :)

File: framework/src/meshgenerators/AdvancedExtruderGenerator.C
Comment (nmnobre) [reply framework/src/meshgenerators/AdvancedExtruderGenerator.C:17]: Yes, that's what I asking above:

> 4. In AdvancedExtruderGenerator::generate() I'm building PRISM21s, so we might need to add formal support for these?

I'll add it to my list and if I have some time I'll do it later on. :)

File: framework/src/meshgenerators/ElementGenerator.C
Comment (roystgnr) [inline framework/src/meshgenerators/ElementGenerator.C:117]: Can we just replace this whole function with `return Elem::build(Utility::string_to_enum<ElemType>(type)).release()`?

Comment (roystgnr) [reply framework/src/meshgenerators/ElementGenerator.C:117]: I'd personally change it to return `unique_ptr` too, but maybe in a separate commit, since that's a change to a public API and I wouldn't want to hold up this PR if some downstream app has started using it.

File: modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C
Comment (roystgnr) [inline modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C:403]: Do we need the switch statement here?  `return mesh.add_elem(libMesh::Elem::build(type));` should cover it, no?  We could test for `Elem::dim() < 3` if we want to keep the protection that `mooseError` gives.

**Review by roystgnr — DISMISSED** [review]
I see only one bug that clearly needs to be fixed; the rest of my comments are either suggestions for quick improvements or questions about future refactoring.

Ah, hell.  That exact same bug afflicts every other `AdvancedExtruderGenerator` case too, doesn't it?  I'll make this an "Approve" rather than "Request changes", then; if you don't want to put in a zillion unrelated fixes here then let me know and I'll put them into a separate PR after this is merged.

File: framework/include/utils/MooseLagrangeHelpers.h
Comment (lindsayad) [reply framework/include/utils/MooseLagrangeHelpers.h:66]: yes if we templated those, then that would eliminate the need for this file

File: framework/src/meshgenerators/AdvancedExtruderGenerator.C
Comment (GiudGiud) [reply framework/src/meshgenerators/AdvancedExtruderGenerator.C:701]: There is a 
```
  mesh->set_isnt_prepared();
```
call at the end of the generator, I think this makes sure we dont rely on those?

Comment (roystgnr) [reply framework/src/meshgenerators/AdvancedExtruderGenerator.C:701]: Sadly, there's no way to not rely on `remote_elem` neighbor links.  There's no non-parallel way to distinguish between a ghosted element that has no `neighbor_ptr` for a side because it really has no neighbor for that side vs one that has no `neighbor_ptr` because it *should* have a pointer to `remote_elem` but doesn't.  There's not even a fast parallel way to do it, which is a bit annoying for Nemesis reads.

Comment (GiudGiud) [reply framework/src/meshgenerators/AdvancedExtruderGenerator.C:701]: ok I misspoke.
I meant arent we recalculating them when we prepare the mesh? We do a find_neighbors

File: modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C
Comment (GiudGiud) [reply modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C:403]: changed in https://github.com/idaholab/moose/pull/25742

File: framework/src/meshgenerators/AdvancedExtruderGenerator.C
Comment (roystgnr) [reply framework/src/meshgenerators/AdvancedExtruderGenerator.C:701]: No, sadly, you didn't misspeak.  I understand that we're getting to `find_neighbors()`, so I'm explaining why `find_neighbors()` doesn't work.  It can determine when a `nullptr` (or even a bad pointer) should be replaced by a pointer to an existing local or ghosted element, but it has no code for determining when a `nullptr` should be replaced by a pointer to `remote_elem`.  It relies on that subset of the neighbor pointers being correct going in to the function (and so it doesn't even check for bad `remote_elem` pointers that should be pointing to a local element instead, like it does for bad local pointers).

It would be possible to write a `find_remote_neighbors()` to get `remote_elem` links fixed up from scratch too, with a `pull_parallel_vector_data` to query every ghosted element's owner for its neighbor ids, but that'd be slow enough that I don't want to write it until we have no other options.  Personally I'd rather move in the other direction some day, even - keep local pointer links always in sync with mesh adaptivity, rather than doing broad sweeps over a whole mesh to fix the relatively small fraction of links that get invalidated; then we'd only call `find_neighbors()` in special cases like right after mesh input from file.

Comment (nmnobre) [reply framework/src/meshgenerators/AdvancedExtruderGenerator.C:701]: Since we are strictly flipping the element upside down, each lateral face should still be defined by the same set of nodes, so each lateral face should still be the same logical side. The bases should swap, but we aren't doing any remote linkage with those, because we don't have to, by the very nature of the extrusion procedure.

Comment (roystgnr) [reply framework/src/meshgenerators/AdvancedExtruderGenerator.C:701]: Oh!  It didn't click with me that we're flipping in zeta.  And that applies to all the other extruded elements too.  Yeah, this is fine.

**Review by roystgnr — DISMISSED** [review]

**Comment by GiudGiud** [comment]
@nmnobre looks like the PR was set to not allow edits from maintainers?
So you'll have to rebase it on next to deal with the conflict (or devel if next has merged already)

**Comment by nmnobre** [comment]
> @nmnobre looks like the PR was set to not allow edits from maintainers?

I'm working on a different fork as I don't have permissions to edit this repository directly, that's why. :)

> So you'll have to rebase it on next to deal with the conflict (or devel if next has merged already)

No worries, I've just done so. I've also realised that I forgot to remove the now unnecessary includes in `PatchSidesetGenerator.C` in #25756. 🙈 To avoid another PR and rebasing yet again, I've added another commit to this PR to remove them. Hope that's okay, even though it's an unrelated change. Sorry... :/

Before merging, can you please comment on point (3) in my initial post above? I assume (1), (2) and (4) are fine?

**Review by GiudGiud — DISMISSED** [review]

**Comment by GiudGiud** [comment]
> Hope that's okay, even though it's an unrelated change. Sorry... :/

yeah that's ok. Small changes are not a burden for the reviewer.

**Comment by roystgnr** [comment]
> Before merging, can you please comment on point (3) in my initial post above? I assume (1), (2) and (4) are fine?

I'm guessing the root cause of (3) is that the MOOSE test framework makes it super easy to add a new regression test, but not so easy (or even possible? I've never seen it, but it'd be a *really* useful feature to have...) to add a test against an analytic/manufactured solution.  So, even when we can set up a problem with a closed-form solution, we tend to set it up as a regression test instead.

I want to mock Guillaume for writing a cubic "quadratic", but I'm currently trying to dig through a bug in an unrelated regression test where I golded an obviously-bad solution two years ago, so who am I to talk.

**Comment by GiudGiud** [comment]
> I want to mock Guillaume for writing a cubic "quadratic",

what where?

**Comment by GiudGiud** [comment]
oh right no this was on purpose.

The point was to not represent it exactly, and a cubic being fitted on a quadratic FE fit the bill

**Comment by nmnobre** [comment]
> The point was to not represent it exactly, and a cubic being fitted on a quadratic FE fit the bill

Right, in that case, should we go quartic now?

**Comment by GiudGiud** [comment]
It's fine as it is. There's some value in testing that the representation remains exact as well.

**Comment by nmnobre** [comment]
> Can I change the name of the functions to cubic_elem, cubic_nodal? So it's clear the order matches for cubic elements but not quadratic elements? Incidentally, I think we can go further, I think we could just have cubic_elem and cubic_nodal functions for all finite element orders and forego the lower order functions, since we effectively cap the order with base_elem and base_nodal and only then do the projection.

Done, let me know if it looks fine. I'll drop the commit otherwise, no problem.

**Comment by GiudGiud** [comment]
Let's remove that last commit. We don't want to regold if we don't have to. 

You can change the name of the IC though.

**Comment by nmnobre** [comment]
> Let's remove that last commit. We don't want to regold if we don't have to.
> 
> You can change the name of the IC though.

Okie dokie, makes sense. :)
Let's just hope the tests pass this time, been acting a little weird lately.

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks for the contribution @nmnobre !

**Pull request merged**
